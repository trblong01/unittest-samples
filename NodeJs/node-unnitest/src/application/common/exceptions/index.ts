import { ZodError } from 'zod';

export class ValidationException extends Error {
  constructor(error: ZodError) {
    super('APPLICATION_VALIDATION_ERROR');
    this.name = 'ValidationException';
    this.errors = error.issues.map((issue) => ({
      path: issue.path.join('.'),
      message: issue.message,
    }));
  }

  errors: { path: string; message: string }[];
}
