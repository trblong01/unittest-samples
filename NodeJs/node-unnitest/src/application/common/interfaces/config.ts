export interface ApplicationConfig {
  env: 'development' | 'production' | 'test';
  logLevel: 'debug' | 'info' | 'warn' | 'error';
  port: number;
}
