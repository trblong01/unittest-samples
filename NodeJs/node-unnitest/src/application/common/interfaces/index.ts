export * from './config';
export * from './logger';
export * from './posts-repository';
