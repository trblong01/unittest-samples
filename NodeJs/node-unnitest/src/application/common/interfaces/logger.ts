type LogMethod = {
  (message: { [key: string]: unknown }): void;
};

export interface Logger {
  debug: LogMethod;
  info: LogMethod;
  warn: LogMethod;
  error: LogMethod;
  fatal: LogMethod;
}
