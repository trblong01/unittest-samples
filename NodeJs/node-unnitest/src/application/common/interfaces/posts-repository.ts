import { Post } from '@domain/entities';

export interface PostsRepository {
  create(parameters: { post: Post }): Promise<{ id: string }>;
  delete(parameters: { id: string }): Promise<void>;
  getById(parameters: { id: string }): Promise<Post | null>;
  list(parameters: { pageNumber: number; pageSize: number }): Promise<{ count: number; posts: Array<Post> }>;
}
