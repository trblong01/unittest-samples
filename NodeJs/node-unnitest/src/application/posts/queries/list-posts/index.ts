export * from './list-posts-query';
