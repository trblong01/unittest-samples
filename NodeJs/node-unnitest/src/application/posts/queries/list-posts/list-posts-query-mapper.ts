import { Post } from '@domain/entities';

export function toDto({
  count,
  pageNumber,
  pageSize,
  posts,
}: {
  count: number;
  pageNumber: number;
  pageSize: number;
  posts: Array<Post>;
}) {
  const totalPages = Math.ceil(count / pageSize);

  return {
    count,
    hasPreviousPage: pageNumber > 1,
    hasNextPage: pageNumber < totalPages,
    pageNumber,
    pageSize,
    posts: posts.map((post) => ({
      id: post.id,
      title: post.title,
    })),
    totalPages,
  };
}
