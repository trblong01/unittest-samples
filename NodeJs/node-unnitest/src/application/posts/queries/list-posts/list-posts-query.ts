import { toDto } from './list-posts-query-mapper';
import { validate } from './list-posts-query-validator';

export type ListPostsQuery = Readonly<{
  pageNumber: number;
  pageSize: number;
}>;

export function makeListPostsQuery({ postsRepository }: Pick<Dependencies, 'postsRepository'>) {
  return async function listPostsQuery(query: ListPostsQuery) {
    await validate(query);

    const { pageNumber, pageSize } = query;

    const { count, posts } = await postsRepository.list({ pageNumber, pageSize });

    return toDto({
      count,
      pageNumber,
      pageSize,
      posts,
    });
  };
}
