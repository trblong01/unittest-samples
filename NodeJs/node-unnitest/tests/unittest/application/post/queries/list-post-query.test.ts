import { ValidationException } from '@application/common/exceptions';
import { PostsRepository } from '@application/common/interfaces';
import { Post } from '@domain/entities';

import { makeListPostsQuery } from '../../../../../src/application/posts/queries/list-posts/list-posts-query';

describe('listPostsQuery', () => {
  function setup() {
    const postsRepository = jest.mocked<Partial<PostsRepository>>({
      list: jest.fn(),
    }) as jest.Mocked<PostsRepository>;

    const listPostQuery = makeListPostsQuery({ postsRepository });

    return {
      listPostQuery,
      postsRepository,
    };
  }

  describe('given an invalid query', () => {
    it('should throw a validation exception', async () => {
      // Arrange
      const { listPostQuery } = setup();

      // Act
      const result = listPostQuery({
        pageNumber: 1,
        pageSize: 1000, // Cannot exceed 50
      });

      // Assert
      await expect(result).rejects.toThrow(ValidationException);
    });
  });

  describe('given a valid query', () => {
    it('should list posts', async () => {
      // Arrange
      const { listPostQuery, postsRepository } = setup();

      postsRepository.list.mockResolvedValueOnce({
        count: 11,
        posts: [
          new Post({
            id: '907b0a0b-9a12-4073-ab27-3ad5927955e9',
            createdAt: new Date(2022, 1, 1),
            title: 'Mock post',
          }),
        ],
      });

      // Act
      const result = await listPostQuery({
        pageNumber: 2,
        pageSize: 10,
      });

      // Assert
      expect(result).toMatchSnapshot();
    });
  });
});
