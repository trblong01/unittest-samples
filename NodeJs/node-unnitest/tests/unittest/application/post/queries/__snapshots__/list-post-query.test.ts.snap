// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listPostsQuery given a valid query should list posts 1`] = `
{
  "count": 11,
  "hasNextPage": false,
  "hasPreviousPage": true,
  "pageNumber": 2,
  "pageSize": 10,
  "posts": [
    {
      "id": "907b0a0b-9a12-4073-ab27-3ad5927955e9",
      "title": "Mock post",
    },
  ],
  "totalPages": 2,
}
`;
